File: src/mutation.js
'use strict';

const { Organism } = require('./organism');

function bitFlipMutation(organism, rng, rate = 0.01) {
  const child = organism.clone();
  for (let i = 0; i < child.size(); i++) {
    if (rng.next() < rate) {
      child.setGene(i, 1 - child.getGene(i));
    }
  }
  return child;
}

function swapMutation(organism, rng, rate = 0.01) {
  const child = organism.clone();
  for (let i = 0; i < child.size(); i++) {
    if (rng.next() < rate) {
      const j = rng.nextInt(0, child.size());
      const tmp = child.getGene(i);
      child.setGene(i, child.getGene(j));
      child.setGene(j, tmp);
    }
  }
  return child;
}

function gaussianMutation(organism, rng, rate = 0.01, sigma = 0.1, min = -Infinity, max = Infinity) {
  const child = organism.clone();
  for (let i = 0; i < child.size(); i++) {
    if (rng.next() < rate) {
      // Box-Muller transform for Gaussian
      const u1 = rng.next() || 1e-10;
      const u2 = rng.next();
      const z = Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2);
      let value = child.getGene(i) + z * sigma;
      value = Math.max(min, Math.min(max, value));
      child.setGene(i, value);
    }
  }
  return child;
}

const STRATEGIES = {
  bitFlip: bitFlipMutation,
  swap: swapMutation,
  gaussian: gaussianMutation,
};

module.exports = {
  bitFlipMutation,
  swapMutation,
  gaussianMutation,
  STRATEGIES,
};

File: python/phyloid_engine/mutation.py
from __future__ import annotations

import math

from .organism import Organism
from .prng import Mulberry32


def bit_flip_mutation(
    organism: Organism, rng: Mulberry32, rate: float = 0.01
) -> Organism:
    """Flip each gene (1-x) with the given probability. Returns a new organism."""
    child = organism.clone()
    for i in range(child.size()):
        if rng.next() < rate:
            child.set_gene(i, 1 - child.get_gene(i))
    return child


def swap_mutation(organism: Organism, rng: Mulberry32, rate: float = 0.01) -> Organism:
    """Swap pairs of genes with the given probability. Returns a new organism."""
    child = organism.clone()
    for i in range(child.size()):
        if rng.next() < rate:
            j = rng.next_int(0, child.size())
            tmp = child.get_gene(i)
            child.set_gene(i, child.get_gene(j))
            child.set_gene(j, tmp)
    return child


def gaussian_mutation(
    organism: Organism,
    rng: Mulberry32,
    rate: float = 0.01,
    sigma: float = 0.1,
    min_val: float = float("-inf"),
    max_val: float = float("inf"),
) -> Organism:
    """Add Gaussian noise to genes, clamped to [min_val, max_val]. Returns a new organism."""
    child = organism.clone()
    for i in range(child.size()):
        if rng.next() < rate:
            u1 = rng.next() or 1e-10
            u2 = rng.next()
            z = math.sqrt(-2 * math.log(u1)) * math.cos(2 * math.pi * u2)
            value = child.get_gene(i) + z * sigma
            value = max(min_val, min(max_val, value))
            child.set_gene(i, value)
    return child
